src/Storages/StorageS3.cpp

```diff
@@ -643,9 +831,23 @@ StorageS3Source::ReaderHolder StorageS3Source::createReader(size_t idx)
     }
     else
     {
-        auto compression_method = chooseCompressionMethod(key_with_info->key, compression_hint);
-        read_buf = createS3ReadBuffer(key_with_info->key, key_with_info->info->size);
+        // LOG_DEBUG(&Poco::Logger::get("Lolchik"), "Lolchik");
+        // LOG_DEBUG(&Poco::Logger::get("Lolchik"), "Lolchik");
+
+        LOG_DEBUG(&Poco::Logger::get("Archive name and size"), "Path: {}, Size: {}", key_with_info->key, key_with_info->info->size);
 
+        auto compression_method = chooseCompressionMethod(key_with_info->key, compression_hint);
+        if (!key_with_info->path_in_archive.has_value())
+        {
+            LOG_DEBUG(&Poco::Logger::get("Perchik"), "Perchik");
+            read_buf = s3_read_buffer_creator->createS3ReadBuffer(key_with_info->key, key_with_info->info->size);
+        }
+        else
+        {
+            LOG_DEBUG(
+                &Poco::Logger::get("Archive structure"), "Path: {}, Name: {}", key_with_info->key, key_with_info->path_in_archive.value());
+            read_buf = archive_holder->getArchiveReader()->readFile(key_with_info->path_in_archive.value(), true);
```

Comment:
In every place where we use `getArchiveReader()->readFile()` we should avoid wrapping buffer into compression codec, because libarchive manages compression by itself. So what happen here:
1. We read `tar.gz` file with libarchive, which will return uncompressed stream
2. We wrap it into `ZlibInflatingReadBuffer` while `read_buf` is already uncompressed:
![image](https://github.com/ClickHouse/ClickHouse/assets/3684697/87e6198c-1820-42fe-91fa-9087e15a14e9)

Reply:
Used CompressionMethod::None when reading from archive